torch/nn/modules/container.py

```diff
@@ -135,6 +135,11 @@ def __add__(self, other) -> 'Sequential':
                              'of Sequential class, but {} is given.'.format(
                                  str(type(other))))
 
+    def pop(self, key: Union[int, slice]) -> Module:
+        v = self[key]
+        del self[key]
```

Comment:
The next keys should be shifted back down no? (Same for list)

Reply:
Hi, @albanD! Thanks for your question.

I think we don't require it. We are using `del` which itself backs down. Pythonic example:
```python
In [33]: l = [1, 2, 3, 4, 5, 6, 7]

In [34]: l
Out[34]: [1, 2, 3, 4, 5, 6, 7]

In [35]: del l[3]

In [36]: l
Out[36]: [1, 2, 3, 5, 6, 7]
```

Reply:
That is what happens on python list. But that is not the case for self.
It will call into the `__delitem__` function above. And as you can see there, it will only remove the entry with that key.

Reply:
Thanks for the inputs, @albanD! But I wonder, how did the test pass? Also, I see in this case the layers back downs by themselves.

<details>
<summary>Script (Click Please)</summary>

<p>

```python
In [1]: import torch

In [2]: from torch import nn

In [3]: l1 = nn.Linear(1, 2)

In [4]: l2 = nn.Linear(2, 3)

In [5]: l3 = nn.Linear(3, 4)

In [6]: l4 = nn.Linear(4, 5)

In [7]: n = nn.Sequential(l1, l2, l3, l4)

In [8]: n
Out[8]: 
Sequential(
  (0): Linear(in_features=1, out_features=2, bias=True)
  (1): Linear(in_features=2, out_features=3, bias=True)
  (2): Linear(in_features=3, out_features=4, bias=True)
  (3): Linear(in_features=4, out_features=5, bias=True)
)

In [9]: n.pop(2)
Out[9]: Linear(in_features=3, out_features=4, bias=True)

In [10]: n
Out[10]: 
Sequential(
  (0): Linear(in_features=1, out_features=2, bias=True)
  (1): Linear(in_features=2, out_features=3, bias=True)
  (3): Linear(in_features=4, out_features=5, bias=True)
)

In [11]: del n[1]

In [12]: n
Out[12]: 
Sequential(
  (0): Linear(in_features=1, out_features=2, bias=True)
  (3): Linear(in_features=4, out_features=5, bias=True)
)
```

</p>
</details>

I think this is because we use `del` in our impl, not `__delitem__`. Added to these, `pop` always returns the index we want to remove.

Could you please let me know if I'm missing something? I still think we don't need to back down the next keys after popping up the elements.

There's an implementation of `pop` for `ModuleDict`. https://github.com/pytorch/pytorch/blob/f5b460b200611b6697d1b21926411f8190c102f3/torch/nn/modules/container.py#L386-L394

Reply:
The reason to shift other keys is because of the index mismatch. In your demonstration, on for input[10],

```
In [10]: n
Out[10]: 
Sequential(
  (0): Linear(in_features=1, out_features=2, bias=True)
  (1): Linear(in_features=2, out_features=3, bias=True)
  (3): Linear(in_features=4, out_features=5, bias=True)
)
```

We see that the `len` for `n` is 3, but there is nothing at index `2`. That is why we need to shift `3` to `2`.

https://github.com/pytorch/pytorch/blob/f5b460b200611b6697d1b21926411f8190c102f3/torch/nn/modules/container.py#L393

Also `del` here dispatches to `__delitem__` above which only removes the key from `self._modules` (but doesn't reorder it)

Reply:
Ooo, right! Now I got. Thanks, @kshitij12345!